libkbfs/keybase_daemon_rpc.go

```diff
@@ -250,64 +272,54 @@ func (k *KeybaseDaemonRPC) AddProtocols(protocols []rpc.Protocol) {
 	if protocols == nil {
 		return
 	}
+
+	k.lock.RLock()
+	defer k.lock.RUnlock()
+
 	if k.protocols != nil {
 		k.protocols = append(k.protocols, protocols...)
 	} else {
 		k.protocols = protocols
 	}
+
+	// If we are already connected, register these protocols.
+	if k.server != nil {
+		for _, p := range protocols {
+			err := k.server.Register(p)
+			if err != nil {
+				k.log.Warning("add protocol %q error", p.Name)
+			}
+		}
+	}
 }
 
 // OnConnect implements the ConnectionHandler interface.
 func (k *KeybaseDaemonRPC) OnConnect(ctx context.Context,
 	conn *rpc.Connection, rawClient rpc.GenericClient,
-	server *rpc.Server) error {
-
-	// Protocols that KBFS requires
-	protocols := []rpc.Protocol{
-		keybase1.LogUiProtocol(daemonLogUI{k.daemonLog}),
-		keybase1.IdentifyUiProtocol(daemonIdentifyUI{k.daemonLog}),
-		keybase1.NotifySessionProtocol(k),
-		keybase1.NotifyKeyfamilyProtocol(k),
-		keybase1.NotifyPaperKeyProtocol(k),
-		keybase1.NotifyFSRequestProtocol(k),
-		keybase1.NotifyTeamProtocol(k),
-		keybase1.TlfKeysProtocol(k),
-		keybase1.ReachabilityProtocol(k),
-		keybase1.ImplicitTeamMigrationProtocol(k),
-	}
-
-	// Add simplefs if set
-	if k.simplefs != nil {
-		protocols = append(protocols, keybase1.SimpleFSProtocol(k.simplefs))
-	}
-
-	// Add git if set
-	if k.gitHandler != nil {
-		protocols = append(protocols, keybase1.KBFSGitProtocol(k.gitHandler))
-	}
-
-	if k.protocols != nil {
-		protocols = append(protocols, k.protocols...)
-	}
-
-	if k.notifyService != nil {
-		k.log.Warning("adding NotifyService protocol")
-		protocols = append(protocols, keybase1.NotifyServiceProtocol(k.notifyService))
-	}
-
-	for _, p := range protocols {
-		err := server.Register(p)
-		if err != nil {
-			if _, ok := err.(rpc.AlreadyRegisteredError); !ok {
-				return err
+	server *rpc.Server) (err error) {
+	err = func() error {
+		k.lock.RLock()
+		defer k.lock.RUnlock()
+
+		for _, p := range k.protocols {
+			err := server.Register(p)
+			if err != nil {
+				if _, ok := err.(rpc.AlreadyRegisteredError); !ok {
+					return err
+				}
```

Comment:
Instead of nested `if` statements here, I'd prefer a type switch.
```go
switch err.(type) {
case nil:
case rpc.AlreadyRegisteredError:
default:
    return err
}
return nil
```